### DISCUSSION TITLE
Image Stitching angle issue

### DISCUSSION THREAD

**Discussion by CodePeacock** [OpenStitching/stitching#166] [Q&A]
Can your package stitch images that are at different angle from each other? Given they are similar. I'm trying to stitch images taken in my home to a 3D panorama. Please respond ASAP 😊

**Comment by lukasalexanderweber** [comment]
can you make a sketch or do you already have images?

**Comment by CodePeacock** [comment]
I have the images ready. I have tried to stitch them but it fails to do so.

**Reply by lukasalexanderweber** [reply]
what have you tried? If you use `stitch`, try `-v` to see the intermediate results, and please share the error message and the resulting matches_graph

**Reply by CodePeacock** [reply]
```
(.venv) X:\work\image-stitching\img>stitch *
stitching 1698049171596.jpg 1698049171630.jpg 1698049171666.jpg 1698049171697.jpg 1698049171734.jpg 1698049596101.jpg 1698049596138.jpg 1698049596179.jpg 1698049596236.jpg 1698049596269.jpg 1698049596299.jpg 1698049596337.jpg into result.jpg
x:\work\image-stitching\.venv\Lib\site-packages\stitching\subsetter.py:32: StitchingWarning: Not all images are included in the final panorama.
                          If this is not intended, use the 'matches_graph_dot_file'
                          parameter to analyze your matches. You might want to
                          lower the 'confidence_threshold' or try another 'detector'
                          to include all your images.
  warnings.warn(
```

> graph matches_graph{
"1698049171596.jpg" -- "1698049171630.jpg"[label="Nm=137, Ni=76, C=1.54786"];
"1698049171596.jpg" -- "1698049171666.jpg"[label="Nm=110, Ni=76, C=1.85366"];
"1698049171596.jpg" -- "1698049171697.jpg"[label="Nm=121, Ni=65, C=1.46727"];
"1698049171596.jpg" -- "1698049171734.jpg"[label="Nm=159, Ni=116, C=2.08259"];
"1698049171666.jpg" -- "1698049596179.jpg"[label="Nm=130, Ni=58, C=1.23404"];
"1698049596101.jpg" -- "1698049596138.jpg"[label="Nm=103, Ni=53, C=1.36247"];
"1698049596101.jpg" -- "1698049596179.jpg"[label="Nm=36, Ni=23, C=1.2234"];
"1698049596236.jpg" -- "1698049596269.jpg"[label="Nm=173, Ni=116, C=1.93656"];
"1698049596236.jpg" -- "1698049596337.jpg"[label="Nm=117, Ni=68, C=1.57773"];
"1698049596299.jpg";
}

**Reply by lukasalexanderweber** [reply]
So it's working? This just say that one image is not included in the final panorama. If you run `stitch * --confidence_threshold 0` you might even be able to get this one included as well

**Reply by CodePeacock** [reply]
I'll try that. But my question still remains. I want to use this to generate a 3d panorama which doesn't warp too much and can be used to showcase a room. Basically I want to generate panorama for a virtual tour of homes.

**Comment by CodePeacock** [comment]
```
(.venv) X:\work\image-stitching\img>stitch * --confidence_threshold 0
stitching 1698049171596.jpg 1698049171630.jpg 1698049171666.jpg 1698049171697.jpg 1698049171734.jpg 1698049596101.jpg 1698049596138.jpg 1698049596179.jpg 1698049596236.jpg 1698049596269.jpg 1698049596299.jpg 1698049596337.jpg result.jpg into result.jpg
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "x:\work\image-stitching\.venv\Scripts\stitch.exe\__main__.py", line 7, in <module>
  File "x:\work\image-stitching\.venv\Lib\site-packages\stitching\cli\stitch.py", line 312, in main
    panorama = stitcher.stitch(images, feature_masks)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "x:\work\image-stitching\.venv\Lib\site-packages\stitching\stitcher.py", line 102, in stitch
    cameras = self.refine_camera_parameters(features, matches, cameras)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "x:\work\image-stitching\.venv\Lib\site-packages\stitching\stitcher.py", line 157, in refine_camera_parameters
    return self.camera_adjuster.adjust(features, matches, cameras)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "x:\work\image-stitching\.venv\Lib\site-packages\stitching\camera_adjuster.py", line 48, in adjust
    raise StitchingError("Camera parameters adjusting failed.")
stitching.stitching_error.StitchingError: Camera parameters adjusting failed.
```

**Reply by CodePeacock** [reply]
Can this discussion be made private so I could share the images, if the images are the issue?

**Reply by lukasalexanderweber** [reply]
without `--confidence_threshold 0`, does the stitching work?

**Reply by CodePeacock** [reply]
No. That's the error after setting to 0.

**Reply by CodePeacock** [reply]
Without it works but the panorama doesn't include all images

**Reply by lukasalexanderweber** [reply]
you could try adding  `--adjuster reproj`, but if this isn't working you need to find the cause by yourself.

**Reply by CodePeacock** [reply]
That didn't work. I guess I'm on my own. Thanks for help @lukasalexanderweber

**Reply by lukasalexanderweber** [reply]
Just another tip for taking Images: stand still and just rotate around yourself

**Comment by CodePeacock** [comment]
I have another query. Does your package use GPU or any GPU that is registered with the device.? Also I've found what was causing that Camera parameters error.

**Reply by lukasalexanderweber** [reply]
https://stackoverflow.com/questions/63601580/use-gpu-with-opencv-python
